From 316d7851500f30b42424e6a8d89e66d4f346f2dd Mon Sep 17 00:00:00 2001
From: Joachim Wiberg <troglobit@gmail.com>
Date: Tue, 7 May 2024 15:41:53 +0200
Subject: [PATCH 2/9] Allow SR_EV_DONE to return any error to sysrepocfg
Organization: Wires

Importing a system configuration with sysrepocfg the model callbacks do
their best to validate the configuration and prepare the transaction in
SR_EV_CHANGE.  In cases when dealing with a complex system, like adding
a route, address, or other change using iproute2, we may still get into
trouble when actually sending off the transaction in SR_EV_DONE.

If something does go wrong we'd like to know this by the exit code of
sysrepocfg, not by inspecting logs or expected system state.

This patch is a clumsy way of forcing the (first) error to bubble up to
the surface and cause a non-zero exit code from sysrepocfg.

Signed-off-by: Joachim Wiberg <troglobit@gmail.com>
---
 src/shm_sub.c | 22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/src/shm_sub.c b/src/shm_sub.c
index 38b189ef..3fe8c8e6 100644
--- a/src/shm_sub.c
+++ b/src/shm_sub.c
@@ -3587,7 +3587,7 @@ sr_shmsub_change_listen_relock(sr_sub_shm_t *sub_shm, sr_lock_mode_t mode, struc
 sr_error_info_t *
 sr_shmsub_change_listen_process_module_events(struct modsub_change_s *change_subs, sr_conn_ctx_t *conn)
 {
-    sr_error_info_t *err_info = NULL;
+    sr_error_info_t *err_info = NULL, *err_tmp = NULL;
     uint32_t i, data_len = 0;
     char *data = NULL, *shm_data_ptr;
     int ret = SR_ERR_OK, filter_valid;
@@ -3609,7 +3609,6 @@ sr_shmsub_change_listen_process_module_events(struct modsub_change_s *change_sub
         }
     }
     if (i == change_subs->sub_count) {
-        /* no new module event */
         goto cleanup;
     }
 
@@ -3715,6 +3714,11 @@ process_event:
                 }
                 break;
             }
+        } else if (sub_info.event == SR_SUB_EV_DONE) {
+            if (ret && err_code == SR_ERR_OK) {
+                /* unexpected, callback actually failed, save for later. */
+                err_code = ret;
+            }
         }
 
         /* subscription processed this event, decrease subscriber count right away for possible later abort event */
@@ -3747,6 +3751,11 @@ process_event:
         }
         break;
     case SR_SUB_EV_DONE:
+        if (err_code) {
+            /* prepare unexepected error from session to be written to SHM */
+            sr_errinfo_new(&err_info, err_code, "Oups, error detected in SR_EV_DONE");
+        }
+        break;
     case SR_SUB_EV_ABORT:
         /* nothing to do */
         break;
@@ -3769,14 +3778,19 @@ process_event:
 
         /* SUB WRITE URGE LOCK */
         if (sr_shmsub_change_listen_relock(sub_shm, SR_LOCK_WRITE_URGE, &sub_info, change_sub, change_subs->module_name,
-                ret, filter_valid, ev_sess, &err_info)) {
+                ret, filter_valid, ev_sess, &err_tmp)) {
+	     if (err_tmp)
+		     err_info = err_tmp;
             goto cleanup;
         }
         sub_lock = SR_LOCK_WRITE_URGE;
 
         /* finish event */
-        err_info = sr_shmsub_listen_write_event(sub_shm, 0, err_code, &shm_data_sub, data, data_len,
+        err_tmp = sr_shmsub_listen_write_event(sub_shm, 0, err_code, &shm_data_sub, data, data_len,
                 change_subs->module_name, err_code ? "fail" : "success");
+        if (err_tmp) {
+            err_info = err_tmp;
+        }
     } else {
         SR_LOG_DBG("EV LISTEN: \"%s\" \"%s\" ID %" PRIu32 " priority %" PRIu32 " success (remaining %" PRIu32 " subscribers).",
                 change_subs->module_name, sr_ev2str(sub_info.event), sub_info.operation_id, sub_info.priority,
